### ISSUE TITLE
Replacement for `expect_args()` and `expect_called()` from `mockery`?

### ISSUE THREAD

**Issue by fh-mthomson** [r-lib/testthat#1889]
I'm experimenting migrating to `testthat` 3e. The transition path from `with_mock()` and `local_mock()` is clear (per deprecation messaging for `testthat::with_mock()` and docs for `?testthat::with_mocked_bindings`_.
> with_mock() and local_mock() are superseded in favour of [with_mocked_bindings()](http://127.0.0.1:14159/help/library/testthat/help/with_mocked_bindings) and [local_mocked_bindings()](http://127.0.0.1:14159/help/library/testthat/help/local_mocked_bindings).

However, I'm still relying on the following from the `mockery` package:
* `mockery::expect_called()`
* `mockery::expect_args()`
* `mockery::mock_args()`
* `mockery::mock()`

Are there `testthat` equivalents for these `mockery` functions? 

PS forgive me if I'm oversimplifying!

**Comment by fh-mthomson** [comment]
At a high-level, testing how a lower-level helper function is called:

- `mockery::expect_called()`: helper function is called (1A and 2A below)

if it _is_ called, the expected arguments....
- `mockery::mock_args()`: ... are passed through (2B below)
- `mockery::expect_args()`:  ... are passed through with the expected values (2C below)

To illustrate (made up examples):
```
library(testthat)

# function definitions ----
parent_function <- function(arg1, arg2) {
  if (arg1 == arg2) {
    arg3 <- ifelse(arg1 > 0, "pos", "neg")

    helper_function(
      arg1 = arg1,
      arg2 = arg2,
      arg3 = arg3
    )
  } else {
    "different"
  }
}

helper_function <- function(arg1, arg2, arg3 = NULL) {
  cli::cli_alert_info("{arg3}!")
  as.character(glue::glue("same: {arg1}=={arg2}"))
}

# behavior ----
parent_function(1, 2)
parent_function(1, 1)
parent_function(-1, -1)

# testing ----
text_same <- "same: 1==1"
text_diff <- "different"

## initial: mockery fns ----

### 1. helper fn NOT called ----
mock_helper <- mockery::mock(text_same)
with_mock(
  "helper_function" = mock_helper,
  result <- parent_function(1, 2)
)

expect_equal(result, text_diff)
mockery::expect_called(mock_helper, 0)

### 2. helper fn IS called ----
mock_helper <- mockery::mock(text_same)
with_mock(
  "helper_function" = mock_helper,
  result <- parent_function(1, 1)
)

expect_equal(result, text_same)

#### 2A: we only care if helper function is called ----
mockery::expect_called(mock_helper, n = 1)

#### 2B: we care about the names of the args passed through to the fn call ----
expect_true("arg3" %in% names(mockery::mock_args(mock_helper)[[1]]))

#### 2C: we care about the values of the args passed through to the helper ----
mockery::expect_args(
  mock_helper,
  n = 1,
  arg1 = 1,
  arg2 = 1,
  arg3 = "pos"
)

## (partially) revised to testthat mocks  ----

# clear: test expected *result* with mocked function (`local_mocked_bindings()` or `with_mocked_bindings()`). e.g., 
local_mocked_bindings(
  helper_function = mock_helper
)

expect_equal(parent_function(1, 2), text_diff)
expect_equal(parent_function(1, 1), text_same)

# unclear: how to test versions B (arg values) and C (arg names) above
```

**Comment by hadley** [comment]
@fh-mthomson does this help?

**Comment by fh-mthomson** [comment]
Thank you so much for the thoughts and suggestions! Sorry for the delayed reply -- this slipped down my priority list (its only called a handful of times in the codebase I've inherited).

The examples I provided are definitely oversimplifying, since the tests are _very_ indirect, thus hard to distill down to a reprex that is "intuitive" for why we'd care about the branching/args in the first place. 

The more I think about it, the more it feels like a code smell wherein we should refactor the tests to be more direct. I don't think they're solving the problem they're meant to. Going to close, at least while I noodle on it further.
